keps/sig-cluster-lifecycle/wgs/20200108-instance-specific-componentconfig.md

```diff
@@ -0,0 +1,644 @@
+  nodefs.available: 0%
+  nodefs.inodesFree: 0%
+evictionPressureTransitionPeriod: 0s
+featureGates:
+  DynamicKubeletConfig: true
+  TaintBasedEviction: true
+healthzBindAddress: 127.0.0.1
+healthzPort: 10248
+kind: KubeletConfiguration
+rotateCertificates: true
+staticPodPath: /etc/kubernetes/manifests
+```
+
+The approach of performing a strategic merge patch of instance-specific over
+shareable config before unmarshalling can decrease the complexity by:
+
+- Using an already proven method from apimachinery.
+- Avoiding complexity around automatic conversions during unmarshaling that might
+  default unset instance-specific fields, which would prevent us from
+  differentiating fields set by the user from fields set by the defaulter.
+- Maintain the independent merge for different fields on each component.
+- The number of steps needed for the strategy merge approach is substantially
+  smaller than the case-by-base solution.
+
+An important note for this solution is the requirement that both sharable and
+instance-specific configuration has the same Kind and APIVersion. This avoids
+the complexities of trying to merge across Kinds, or perform conversions before
+merging.
+
+### Code merge example
+
+In the example above, it's possible to control the patch strategy used in the
+configuration struct field tagging:
+
+```golang
+// External Kubelet configuration API
+// k8s.io/kubelet/config/v1beta1/types.go
+
+type KubeletConfiguration struct {
+    ClusterDNS []string `json:"clusterDNS" strategyMerge:"replace"`
+    FeatureGates map[string]bool `json:"featureGates,omitempty" strategyMerge:"merge"`
+}
+```
+
+The StrategicMergePatch operates directly on the YAML but uses the strategies
+specified in struct tags on the passed-in object. An empty configuration object
+can be passed to the patch function to specify the desired patch strategies.
+
+```golang
+// Some general configuration utilities
+// utils/patch.go
+...
+import (
+    "sigs.k8s.io/yaml"
+)
+...
+
+// ConvertYAMLToJSON convert the YAML configuration to JSON.
+func ConvertYAMLToJSON(yamlSharable, yamlInstance []byte) ([]byte, []byte, error) {
+    jsonSharable, err := yaml.YAMLToJSON(yamlSharable)
+    If err != nil {
+        return nil, nil, err
+    }
+
+jsonInstance, err := yaml.YAMLToJSON(yamlInstance)
+    If err != nil {
+        return nil, nil, err
+    }
+
+    return jsonSharable, jsonInstance, nil
+}
+
+// PatchConfiguration applies a strategic merge patch and returns the patched JSON result,
+// it is necessary to convert both configurations from YAML to JSON before applying the merge.
+func PatchConfiguration(jsonSharable, jsonInstance []byte) []byte, error {
+    obj := &kubeletv1beta1.KubeletConfiguration{}
+
+    patched, err := strategicpatch.StrategicMergePatch(jsonSharable, jsonInstance, obj)
+    if err != nil {
+        return []byte{}, err
+    }
+
+    return patched, err
+}
+```
+
+The steps for this merges is simpler than the alternative approach:
```

Comment:
I like this, and think it's better than the other approach above. But I don't understand fully is what the now-updated flag strategy is?

Is it: SharedConfig -> InstanceConfig (patch) -> Flags (patch) -> CompleteConfig
Or SharedConfig -> (InstanceConfig OR Flags) (patch) -> CompleteConfig
Or SharedConfig -> InstanceConfig (patch) -> CompleteConfig?

i.e. if we treat flags essentially as a JSON patch (by applying them on an object & encoding), it can be transparent to the strategic-merge-patch applying where the patch came from (flags or file). Thus, we could remove 90%-ish of the flags, that are covered by the InstanceConfig flow, and just keep a couple "obvious" ones, like `--hostname-override` for kube-proxy.

I'm of the opinion that we should keep a small set of flags, making the the custom file with the init container JUST to do the hostname override is not user-friendly IMO.

Reply:
For kube-proxy, if we can make flags and config mutually exclusive overall I think that's a good simplifying approach (use flags OR `--config` and `--instance-config`). I believe this is already how it works in kube-proxy for "shared" config fields, solving instance-specific config could let us extend that to all config.

Kubelet is stuck in the flag-merging world until we can get its ComponentConfig to GA and deprecate the flags, since it's already that way in Beta.

I'm not sure it buys us much to decode flags on an object, then encode it and use it as a patch; we're already patching by decoding flags on the object in the first place. Probably simpler to just have the two exclusive paths depending on if `--config` is specified. 

> I'm of the opinion that we should keep a small set of flags, making the the custom file with the init container JUST to do the hostname override is not user-friendly IMO.

I agree that part is still heavyweight. @neolit123 wrote a doc on other possible CLI options a while back: https://docs.google.com/document/d/1Dvct469xfjkgy3tjWMAKvRAJo4CmGH4cgSVGTDpay6A/edit. I still think a way to patch on the command line could be a good idea for future UX improvement, but it's out of scope for this KEP IMO.